reconcile/openshift_prometheusrules.py

```diff
@@ -0,0 +1,319 @@
+        f.close()
+        return libsonnet_file_path
+    except Exception as e:
+        raise Exception("Template {} not found {}"
+                        .format(LIBSONNET_TEMPLATE_FILE, e)
+                        )
+
+
+def load_templates():
+    """Loads templates from the Python package"""
+    # Todo we should move this to common utils
+    package_templates_dir = os.path.dirname(templates.__file__)
+    template_loader = jinja2.FileSystemLoader(
+      searchpath=package_templates_dir
+      )
+    return jinja2.Environment(loader=template_loader)
+
+
+def fetch_desired_state(performance_parameters, ri):
+    jb_client = init_jb_client()
+    defer(lambda: jb_client.cleanup())
+    temp_dir_path = jb_client.get_dir_path()
+    template_env = load_templates()
+
+    # Iterate over all available Performance Parameters
+    for pp in performance_parameters['performance_parameters_v1']:
+        for namespace in pp['app']['namespaces']:
+            if namespace['cluster']['observabilityNamespace'] is not None:
+
+                libsonnet_file_path = generate_libsonnet_file(
+                  pp,
+                  namespace,
+                  temp_dir_path,
+                  template_env
+                  )
+
+                jsonnet_file_path = generate_jsonnet_file(
+                  pp,
+                  namespace,
+                  libsonnet_file_path,
+                  temp_dir_path,
+                  template_env
+                  )
+
+                # Generate the manifests
+                output = jsonnet.generate(jsonnet_file_path, temp_dir_path)
+                resource_name = output['metadata']['name']
+
+                openshift_resource = OR(
+                  output,
+                  QONTRACT_INTEGRATION,
+                  QONTRACT_INTEGRATION_VERSION,
+                  error_details=resource_name
+                  )
+
+                cluster = namespace['cluster']
+                cluster_observability_namespace = cluster.get(
+                  'observabilityNamespace'
+                  )
+
+                ri.add_desired(
+                    cluster_observability_namespace['cluster']['name'],
+                    cluster_observability_namespace['name'],
+                    'PrometheusRule',
+                    resource_name,
+                    openshift_resource
+                )
+
+
+def init_jb_client():
+    jsonnetfile = """{
+        "dependencies": [
+            {
+                "name": "slo-libsonnet",
+                "source": {
+                    "git": {
+                        "remote":
+                        "https://github.com/metalmatze/slo-libsonnet",
+                        "subdir": "slo-libsonnet"
+                    }
+                },
+                "version": "037525d64e7ffe3198acfbaa99482cffced8e9c0"
+            }
+        ]
+    }
+    """
+    return JsonnetBundler(jsonnetfile)
+
+
+def get_observability_namespaces(performance_parameters):
```

Comment:
since we are iterating over `pp` and `observability_namespaces` is initialized each time, this function will only return the result of the last iteration.
if we expect a single item, this should be documented and not run in a loop.
i.e. use `performance_parameters['performance_parameters_v1'][0]` or something similar.

Reply:
Nope, its expected to return all observability namespaces. We're not running this in a loop.

Reply:
what i meant in -
in line 202, we are initializing observability_namespaces for each pp, and eventually returing it. it means that we are only returning the observability_namespaces found in the last iteration. is this intentional?